source/plugins/go/src/ingestion_token_utils.go

```diff
@@ -0,0 +1,437 @@
+						ContainerNodeInventoryBlob      string `json:"CONTAINER_NODE_INVENTORY_BLOB"`
+						KubeEventsBlob                  string `json:"KUBE_EVENTS_BLOB"`
+						KubeHealthBlob                  string `json:"KUBE_HEALTH_BLOB"`
+						KubeMonAgentEventsBlob          string `json:"KUBE_MON_AGENT_EVENTS_BLOB"`
+						KubeNodeInventoryBlob           string `json:"KUBE_NODE_INVENTORY_BLOB"`
+						KubePodInventoryBlob            string `json:"KUBE_POD_INVENTORY_BLOB"`
+						KubePvInventoryBlob             string `json:"KUBE_PV_INVENTORY_BLOB"`
+						KubeServicesBlob                string `json:"KUBE_SERVICES_BLOB"`
+						InsightsMetricsBlob             string `json:"INSIGHTS_METRICS_BLOB"`
+					} `json:"outputStreams"`
+				} `json:"ContainerInsights"`
+			} `json:"extensionConfigurations"`
+		} `json:"content"`
+	} `json:"configurations"`
+}
+
+type IngestionTokenResponse struct {
+	Configurationid    string `json:"configurationId"`
+	Ingestionauthtoken string `json:"ingestionAuthToken"`
+}
+
+func getAccessTokenFromIMDS() (string, int64, error) {
+	Log("Info getAccessTokenFromIMDS: start")
+	useIMDSTokenProxyEndPoint := os.Getenv("USE_IMDS_TOKEN_PROXY_END_POINT")
+	imdsAccessToken := ""
+	var responseBytes []byte
+	var err error
+
+	if (useIMDSTokenProxyEndPoint != "" && strings.Compare(strings.ToLower(useIMDSTokenProxyEndPoint), "true") == 0) {
+		Log("Info Reading IMDS Access Token from IMDS Token proxy endpoint")
+		MCS_ENDPOINT := os.Getenv("MCS_ENDPOINT")
+		var msi_endpoint *url.URL
+		msi_endpoint, err := url.Parse("http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://" + MCS_ENDPOINT + "/")
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating IMDS endpoint URL: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req, err := http.NewRequest("GET", msi_endpoint.String(), nil)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating HTTP request: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req.Header.Add("Metadata", "true")
+
+		//IMDS endpoint nonroutable endpoint and requests doesnt go through proxy hence using separate client
+		httpClient := &http.Client{}
+
+		// Call managed services for Azure resources token endpoint
+		var resp *http.Response = nil
+		for retryCount := 0; retryCount < 3; retryCount++ {
+			resp, err = httpClient.Do(req)
+			if err != nil {
+				message := fmt.Sprintf("getAccessTokenFromIMDS: Error calling token endpoint: %s, retryCount: %d", err.Error(), retryCount)
+				Log(message)
+				SendException(message) // send the exception here because this error is not returned. The calling function will send any returned errors to telemetry.
+				continue
+			}
+
+			if resp != nil && resp.Body != nil {
+			  defer resp.Body.Close()
+			}
+
+			Log("getAccessTokenFromIMDS: IMDS Response Status: %d, retryCount: %d", resp.StatusCode, retryCount)
+			if resp.StatusCode != 200 {
```

Comment:
Why are we retrying if its any non 200? This leads to retries in case of 400s too.

Reply:
good catch. Fixed .